routes/stake.go

```diff
@@ -0,0 +1,417 @@
+package routes
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"github.com/deso-protocol/core/lib"
+	"github.com/holiman/uint256"
+	"io"
+	"net/http"
+)
+
+type StakeRewardMethod string
+
+const (
+	PayToBalance StakeRewardMethod = "PAY_TO_BALANCE"
+	Restake      StakeRewardMethod = "RESTAKE"
+)
+
+type StakeRequest struct {
+	TransactorPublicKeyBase58Check string            `safeForLogging:"true"`
+	ValidatorPublicKeyBase58Check  string            `safeForLogging:"true"`
+	RewardMethod                   StakeRewardMethod `safeForLogging:"true"`
+	StakeAmountNanos               *uint256.Int      `safeForLogging:"true"`
+	ExtraData                      map[string]string `safeForLogging:"true"`
+	MinFeeRateNanosPerKB           uint64            `safeForLogging:"true"`
+	TransactionFees                []TransactionFee  `safeForLogging:"true"`
+}
+
+type UnstakeRequest struct {
+	TransactorPublicKeyBase58Check string            `safeForLogging:"true"`
+	ValidatorPublicKeyBase58Check  string            `safeForLogging:"true"`
+	UnstakeAmountNanos             *uint256.Int      `safeForLogging:"true"`
+	ExtraData                      map[string]string `safeForLogging:"true"`
+	MinFeeRateNanosPerKB           uint64            `safeForLogging:"true"`
+	TransactionFees                []TransactionFee  `safeForLogging:"true"`
+}
+
+type UnlockStakeRequest struct {
+	TransactorPublicKeyBase58Check string            `safeForLogging:"true"`
+	ValidatorPublicKeyBase58Check  string            `safeForLogging:"true"`
+	StartEpochNumber               uint64            `safeForLogging:"true"`
+	EndEpochNumber                 uint64            `safeForLogging:"true"`
+	ExtraData                      map[string]string `safeForLogging:"true"`
+	MinFeeRateNanosPerKB           uint64            `safeForLogging:"true"`
+	TransactionFees                []TransactionFee  `safeForLogging:"true"`
+}
+
+type StakeTxnResponse struct {
+	SpendAmountNanos  uint64
+	TotalInputNanos   uint64
+	ChangeAmountNanos uint64
```

Comment:
Just for my own understand, do we really need to return inputs and change amount if this txn type can only be active after the balance model fork height?

Reply:
Well we do need the spend amount for identity's UI. You can include additional outputs in a transaction and this field helps show the total amount that will be spent in the transaction. Additionally, having the consistent output makes a lot of our code in deso-js easier to manage. Ideally we deprecate things we don't need eventually, but for now, it's easiest to keep things consistent.